File: python/queue/src/queue.py
class QueueInterface:
    def enqueue(self, item):
        raise NotImplementedError

    def dequeue(self):
        raise NotImplementedError

    def peek(self):
        raise NotImplementedError

    def is_empty(self):
        raise NotImplementedError

    def is_full(self):
        raise NotImplementedError

    def print_queue(self):
        raise NotImplementedError

class Queue(QueueInterface):
    def __init__(self, capacity=5):
        self.items = [None] * capacity
        self.capacity = capacity
        self.size = 0
        self.front = 0
        self.rear = -1

    def enqueue(self, item):
        if self.is_full():
            print("Queue overflow")
            return

        self.rear = (self.rear + 1) % self.capacity
        self.items[self.rear] = item
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            print("Queue underflow")
            return None

        item = self.items[self.front]
        self.items[self.front] = None
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return item

    def peek(self):
        if self.is_empty():
            return None
        return self.items[self.front]

    def is_empty(self):
        return self.size == 0

    def is_full(self):
        return self.size == self.capacity

    def print_queue(self):
        print([self.items[i] for i in range(self.capacity) if self.items[i] is not None])

File: scala/queue/src/main/scala/Queue.scala
trait Queue[T] {
  def enqueue(item: T): Unit
  def dequeue(): Option[T]
  def peek(): Option[T]
  def isEmpty: Boolean
  def isFull: Boolean
  def printQueue(): Unit
}

object Queue {
  private class ArrayQueue[T](capacity: Int) extends Queue[T] {
    private val items: Array[Any] = new Array[Any](capacity)
    private var front: Int = 0
    private var rear: Int = -1
    private var size: Int = 0

    override def enqueue(item: T): Unit = {
      if (isFull) {
        println("Queue overflow")
      } else {
        rear = (rear + 1) % capacity
        items(rear) = item
        size += 1
      }
    }

    override def dequeue(): Option[T] = {
      if (isEmpty) {
        println("Queue underflow")
        None
      } else {
        val item = items(front).asInstanceOf[T]
        front = (front + 1) % capacity
        size -= 1
        Some(item)
      }
    }

    override def peek(): Option[T] = {
      if (isEmpty) None
      else Some(items(front).asInstanceOf[T])
    }

    override def isEmpty: Boolean = size == 0

    override def isFull: Boolean = size == capacity

    override def printQueue(): Unit = {
      print("Queue: [")
      var i = front
      for (_ <- 0 until size) {
        print(items(i))
        if (i != rear) print(", ")
        i = (i + 1) % capacity
      }
      println("]")
    }
  }

  def apply[T](capacity: Int): Queue[T] = new ArrayQueue[T](capacity)
}